Shell

hughleat opened the pull request:
Some changes to the manual_env.

Allows command completion. 
Adds an undoable stack of actions.
Adds a couple of simple search techniques.
Adds a few miscellaneous capabilities.

There are some issues that are going to be or have been fixed in the mainline.
1. After downloading a new dataset the benchmarks do not update without restarting
2. The benchmark is not set from the command line

hughleat committed:
Adding a new manual environment shell

(There were a load of commits in another branch, but git was behaving badly. Oh, well.)

hughleat committed:
Adding a new manual environment shell

(There were a load of commits in another branch, but git was behaving badly. Oh, well.)

hughleat committed:
Minor formatting

hughleat committed:
Minor formatting

hughleat committed:
Minor documentation

hughleat committed:
Minor documentation

hughleat committed:
Removing some dead comments

hughleat committed:
Removing some dead comments

hughleat commented:
Oh, one more thing. When I set the default reward or observation I get this:

/Users/hleather/opt/miniconda3/lib/python3.8/site-packages/compiler_gym-0.1.2-py3.8.egg/compiler_gym/envs/compiler_env.py:503: UserWarning: Changing eager observation space has no effect until reset() is called.
  warnings.warn(
Observation  in 3.7ms

Not sure if that warning is necessary or if there is better way for me to use it.

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+            print("No default reward set. Call set_default_reward")
+            return
+
+        try:
+            num_steps = max(1, int(arg))
+        except ValueError:
+            num_steps = 1
```
Could we add a `greedy -` that runs greedy until no action is selected?

hughleat replied:
Gonna leave that one for the moment. Can we leave it to a feature req (I'm a bit desperate to get back to the deep DF stuff).

ChrisCummins replied:
Yep sounds good!

ChrisCummins replied:
Follow up in #75

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+                    print(f"Step: {i+1} Selected no action")
+                    break
+
+        print(f"Greedy {num_steps} steps in {timer}")
```
If `i != num_steps` it would be nice to print a message to explain why. Perhaps something like "Greedy search for 10 steps stopped after step 5 selected no action"?

hughleat replied:
Done.

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+
+    def do_try_all_actions(self, args):
+        """Tries all actions from this position and reports the results in sorted order by reward"""
+        with Timer() as timer:
+            items = self.get_action_rewards()
+        print(f"Got actions in {timer}")
```
FYI (no action needed) there is a shortcut for this:

```suggestion
        with Timer("Got actions") as timer:
            items = self.get_action_rewards()
```

hughleat replied:
Good one. I've changed those that can be changed.

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+    def __init__(self,
+        action_name,
+        action_index,
+        eager_observation,
+        eager_reward,
```
nitpick: I'd just call these `observation`/`reward` to keep in line with regular gym terminology. We'll drop the "eager" branding soon

same applies for other uses of `eager_` in this file

hughleat replied:
Cool. Wasn't sure if the others had changed.

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+            with Timer(f"Reward {arg}"):
+                print(f"{self.env.reward[arg]:.6f}")
+        else:
+            print("Unknown reward, '" + arg + "'")
```
Consider calling `reward` with no arg to print the default reward, if set. At the moment it errors:
```
compilergym:cBench-v0/dijkstra> reward
Unknown reward, ''
Rewards are listed with command, list_rewards
```

instead:

```
compilergym:cBench-v0/dijkstra> reward
0.000000
Reward IrInstructionCount in 667.8us
```

Same for `observation` with no arg

hughleat replied:
Done

ChrisCummins replied:
Thanks!

ChrisCummins commented on compiler_gym/bin/manual_env.py:
```diff
@@ -16,142 +16,590 @@
+            if check_rewards and eager_reward != old_hist.eager_reward:
+                print(f"Warning previous eager reward at {i}: {hist.action_name} was {hist.eager_reward:.6f} now {eager_reward:.6f}")
+
+    def do_hill_climb(self, arg):
+        """Do some steps of hill climbing.
+            A random action is taken, but only accepted if it has a positive reward.
+            An argument, if given, should be the number of steps to take.
+            The search will try to improve the default reward. Please call set_default_reward if needed.
+        """
```
Not urgent, but it would be nice if hill climbing printed a summary at the end, something like:

`Hill climb completed in 2.12s. Accepted 4 of 100 steps for a total reward of 0.123`

hughleat replied:
Done

ChrisCummins reviewed (changes requested):
This is great, thank you! I love the tab completion and the new stack/search features. Some minor code nitpicks inline.

My biggest request is for more documentation. The old [manual_env](https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env) doesn't have much to say as it can't do much, but your new version is useful enough that it deserves a much more verbose description, maybe even a mini-tutorial. Could you elaborate the module docstring to add something like this?

A welcome message on startup that points the user towards `help` may also be good, presently it drops the user into a commandline without much fanfare:

```
$ python -m compiler_gym.bin.manual_env --env=llvm-ic-v0
Initialized environment in 125.7ms
compilergym:NO-BENCHMARK>
```

BTW, since I've never used the cmd.Cmd class, is it easy to test? Could we have a smoke test that simulates the user selecting some known benchmark and running say a single action?

Cheers,
Chris

hughleat committed:
Bunch of changes request on latest PR.

hughleat committed:
Bunch of changes request on latest PR.

hughleat committed:
Tests

hughleat committed:
Tests

hughleat committed:
Formatting and fixes for pre-commit checks

hughleat committed:
Formatting and fixes for pre-commit checks

ChrisCummins committed:
Verbose failure message for test.

ChrisCummins committed:
Verbose failure message for test.

ChrisCummins commented:
Great, thanks for the changes @hughleat! Love the tests :)

One tiny thing:
```
compilergym:NO-BENCHMARK> tutorial
*** Unknown syntax: tutorial
```

I don't think `tutorial` works. In fact, could you just move the tutorial text into the module docstring? That way it'll get picked up and rendered in the reference docs: https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env

ChrisCummins commented:
BTW the CI tests were failing. Not sure why, as I can reproduce locally just fine. I've added a small patch to make it easier to debug 8d642118e1bdfcb6449a2a9133df1bbfa8686e65

hughleat committed:
Merge branch 'development' of https://github.com/facebookresearch/CompilerGym into shell

hughleat committed:
Merge branch 'shell' of https://github.com/facebookresearch/CompilerGym into shell

ChrisCummins committed:
Remove fixme note.

Fixed in b2592ab076c89ebc33227356d26d42065d2364a6.

hughleat committed:
Fix non-determinism in tests

The behaviour on different machines was different. Possibly due to different
behaviour of the random number generator (even after setting the seed). So, now
all that kind of thing should have been rubbed out.

hughleat committed:
Merge branch 'shell' of https://github.com/facebookresearch/CompilerGym into shell

# Conflicts:
#	tests/bin/manual_env_test.py

hughleat commented:
> I don't think `tutorial` works. In fact, could you just move the tutorial text into the module docstring? That way it'll get picked up and rendered in the reference docs: https://facebookresearch.github.io/CompilerGym/cli.html#module-compiler_gym.bin.manual_env

Changed it from `help tutorial` to `tutorial`. Now extracts from `__doc__`.

hughleat committed:
Moving tutorial to __doc__

And changing `help tutorial` into `tutorial`.

ChrisCummins committed:
[tests] Don't rely on test execution order.

When `make install-test` is used, tests are executed in parallel and
test order is undefined. This adds a new `cBench_dataset` test fixture
that ensures that tests have the cBench dataset available to them.

ChrisCummins reviewed (approved):
Thanks for the updates @hughleat! I added a small fix for a test-execution-order bug 7e88bc763bc39bdf070ed88f6bbf2c80045dbcff.

LGTM, please merge when you're happy :)
